File: norm_linearization/by_taylor_using_julia.jl
# Following  https://juliaintervals.github.io/TaylorModels.jl/latest/

using TaylorModels
using Plots

order = 4
c = 4.0
f(x) = sin(x)
# f(x) = 1 / (1 + exp(-x))
f(x) = x^3 + x^2 + x + 1
g(x) = (f(x) - f(c))^2


# Define domain and expansion point
a = c-5.0 .. c+5.0

# Create Taylor models
tm = TaylorModel1(order, interval(c), a)

# Compute Taylor models of g(x)
ftm = g(tm)

# Plotting
gr()
plot(range(inf(a), stop=sup(a), length=100), x->g(x), 
    label="true", lw=2, xaxis="x", yaxis="||f(x)-f(c)||^2", 
    title="f(x) = x^3 + x^2 + x + 1, order=3, c= 4.0")
plot!(ftm, label="taylor model")



#=
# Define the partition of [-5,5] into 10 intervals
num_partitions = 10
intervals = [(-5 + i * 10 / num_partitions) .. (-5 + (i + 1) * 10 / num_partitions) for i in 0:num_partitions-1]

# Create plot
gr()
plot(title="f(x) = sin(x)", xaxis="x", yaxis="||f(x)-f(c)||^2", legend=:topright)

for a in intervals
    # Create Taylor model
    tm = TaylorModel1(order, interval(mid(a)), a)
    
    # Compute Taylor model of g(x)
    ftm = g(tm)
    
    # Plot Taylor model
    plot!(ftm, label=nothing)  # Remove label to avoid clutter
end

# Plot the true function for reference
x_vals = range(-5, stop=5, length=200)
plot!(x_vals, x -> g(x), label="true", lw=2)
=#

File: sandbox/norm_linearization/by_taylor_using_julia.py
#import juliacall
# jl = juliacall.newmodule("julias_playground")

from juliacall import Main
import numpy as np
import matplotlib.pyplot as plt
import torch

"""
Implementation of example in Taylor Models documentation via Python
"""


if __name__ == '__main__':
    # Compute Taylor Models in Julia
    Main.seval("""
    using TaylorModels

    c = 2.0
    f(x) = 1 / (1 + exp(-x)) # sigmoid
    g(x) = (f(x) - f(c))^2

    # Define domain and expansion point
    a = 1 .. 2.5

    # Create Taylor models
    tm = TaylorModel1(2, interval(c), a)

    # Compute Taylor models of g(x)
    ftm = g(tm)

    coeffs = [[inf(coef), sup(coef)] for coef in ftm.pol.coeffs]  # coefficient bounds
    rem = [inf(ftm.rem), sup(ftm.rem)]  # remainder bound
    """)

    print(Main.ftm)
    c = torch.tensor(Main.c)

    # Compute Values for Plotting
    X = torch.linspace(1., 2.5, 1000)
    X_plot = torch.norm(X - c, p=2, dim=-1).pow(2)

    # Evaluate f(x) for real numbers in Python
    def f(x):
        return 1 / (1 + np.exp(-x))

    def g(x):
        return (f(x) - f(c))**2

    class Polynomial:
        def __init__(self, coeffs: torch.Tensor, rem: torch.Tensor):
            self.coeffs = coeffs
            self.rem = rem

        def approx(self, x):
            return torch.polynomial.polynomialval(x, self.coeffs)

        def bound(self, x):
            y = self.approx(x)
            return y + self.rem[..., 0], y + self.rem[..., 1]

    Y = g(X)

    polynomial = Polynomial(coeffs=torch.tensor(Main.coeffs).mean(-1).flip(dims=(-1,)), rem=torch.tensor(Main.rem))

    Y_approx = polynomial.approx(X)
    Y_lower, Y_upper = polynomial.bound(X)

    plt.figure(figsize=(8, 5))
    plt.plot(X_plot, Y, label="g(x)", linewidth=2, color="black")
    plt.plot(X_plot, Y_approx, label="approx", color="blue")
    plt.plot(X_plot, Y_lower, label="lower bound", color="red")
    plt.plot(X_plot, Y_upper, label="upper bound", color="green")
    plt.xlabel("x")
    plt.ylabel("f(x)")
    plt.legend()
    plt.grid()
    plt.show()

